ModalContainer provides SavedStateRegistries to each of its layers.

rjrjr opened the pull request:
Without this `BackStackContainer` is the only thing that can host `ComposeView` in `Dialog` windows.

This fixes #469, and also fixes #470.

Co-authored-by: Ray Ryan <ray@squareup.com>

rjrjr commented:
This sometimes fails when opening a modal (on tablet, at least). e.g., run the Tic Tac Toe sample, complete a game, and click Exit. I suspect it's masked in cases where the modal is shown immediately at launch.

```
E/TicTacToeComponent: java.lang.IllegalStateException: You can consumeRestoredStateForKey only after super.onCreate of corresponding component
        at androidx.savedstate.SavedStateRegistry.consumeRestoredStateForKey(SavedStateRegistry.java:77)
        at com.squareup.workflow1.ui.androidx.StateRegistryAggregator$lifecycleObserver$1.onStateChanged(StateRegistryAggregator.kt:89)
        at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.java:354)
        at androidx.lifecycle.LifecycleRegistry.addObserver(LifecycleRegistry.java:196)
        at com.squareup.workflow1.ui.androidx.StateRegistryAggregator.attachToParentRegistry(StateRegistryAggregator.kt:145)
        at com.squareup.workflow1.ui.backstack.ViewStateCache.attachToParentRegistry(ViewStateCache.kt:149)
        at com.squareup.workflow1.ui.backstack.BackStackContainer.onAttachedToWindow(BackStackContainer.kt:183)
        at android.view.View.dispatchAttachedToWindow(View.java:17381)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3319)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3326)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3326)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3326)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:1658)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1386)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6733)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:911)
        at android.view.Choreographer.doCallbacks(Choreographer.java:723)
        at android.view.Choreographer.doFrame(Choreographer.java:658)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:897)
        at android.os.Handler.handleCallback(Handler.java:789)
        at android.os.Handler.dispatchMessage(Handler.java:98)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6541)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
```

rjrjr commented on samples/containers/android/src/main/java/com/squareup/sample/container/overviewdetail/OverviewDetailContainer.kt:
```diff
@@ -53,11 +53,8 @@ class OverviewDetailContainer(view: View) : LayoutRunner<OverviewDetailScreen> {
-        .withBackStackStateKeyPrefix(OverviewBackStackKey) + (OverviewDetailConfig to Overview)
-      overviewStub!!.update(rendering.overviewRendering, overviewViewEnvironment)
+      val overviewViewEnvironment = viewEnvironment + (OverviewDetailConfig to Overview)
+      overviewStub!!.update(Named(rendering.overviewRendering, "overview"), overviewViewEnvironment)
```
Explain this.

rjrjr commented:
This is green (enough), but I'm still proving it out against our massive app. Hoping to open it up for review tomorrow.

rjrjr commented on workflow-ui/container-android/src/main/java/com/squareup/workflow1/ui/modal/ModalContainer.kt:
```diff
@@ -48,24 +52,46 @@ public abstract class ModalContainer<ModalRenderingT : Any> @JvmOverloads constr
+    },
+    onRestored = { aggregator ->
+      dialogs.forEach { ref ->
+        // We're only allowed to restore from an INITIALIZED state, but this callback can also be
```
We have copy/pasted this fix in both places we use the aggregator. Can we build it in? I think we'd have to change it to depend on `KeyedStateRegistryOwner` directly. Maybe that's fine? Or vice versa, make the aggregator an implementation detail of `KSRO`.

rjrjr commented on workflow-ui/container-android/src/main/java/com/squareup/workflow1/ui/modal/ModalContainer.kt:
```diff
@@ -75,6 +101,12 @@ public abstract class ModalContainer<ModalRenderingT : Any> @JvmOverloads constr
+            // don't clash with other layers.
+            ref.stateRegistryOwner = KeyedStateRegistryOwner.installAsSavedStateRegistryOwnerOn(
+              view = dialogView,
+              key = Compatible.keyFor(newScreen, i.toString())
```
Can this be built into `createStateRegistryKeyForContainer`? Add an optional `index: Int? = null` parameter, e.g.

rjrjr commented on workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt:
```diff
@@ -82,9 +82,9 @@ internal class StateRegistryAggregator(
-        parentRegistryOwner!!.savedStateRegistry.consumeRestoredStateForKey(parentKey!!)
-      restoreFromBundle(restoredState)
+      parentRegistryOwner!!.savedStateRegistry.takeIf { it.isRestored }?.let {
+        restoreFromBundle(it.consumeRestoredStateForKey(parentKey!!))
```
Probably should always call `restoreFromBundle` here, with `null` if necessary, so that its prune call always happens.

rjrjr commented on workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/WorkflowAndroidXSupport.kt:
```diff
@@ -34,10 +39,52 @@ public object WorkflowAndroidXSupport {
    */
   @WorkflowUiExperimentalApi
-  public fun stateRegistryOwnerFromViewTreeOrContext(view: View): SavedStateRegistryOwner? =
+  public fun requireStateRegistryOwnerFromViewTreeOrContext(view: View): SavedStateRegistryOwner =
```
Get rid of `require` prefix.

vgonda commented on workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt:
```diff
@@ -0,0 +1,300 @@
+      throw IllegalArgumentException("$key is already in use, it cannot be used register $view")
+    }
+    ViewTreeSavedStateRegistryOwner.get(view)?.let {
+      "$view already has ViewTreeSavedStateRegistryOwner: $it"
```
Did you mean to throw this? It looks like a floating, unused string right now.

vgonda commented on workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt:
```diff
@@ -0,0 +1,300 @@
+
+  private fun saveToBundle() = Bundle().apply {
+    doIfRestored { states ->
+      // onWillSave(this@StateRegistryAggregator)
```
I'm not sure what this comment is supposed to communicate. Similar to `// onRestored(this)` below.

rjrjr replied:
Oops! I was experimenting w/whether I could delete those params. I did, meant to delete this.

vgonda reviewed (commented):
I have no arguments about any of this, but I also don't feel I understand it enough to 👍 

If you end up setting up a meeting, I'd love to join (if I can be pinned down. I'm traveling this week, so I know that's a challenge).

rjrjr commented on workflow-ui/container-android/src/main/java/com/squareup/workflow1/ui/backstack/ViewStateCache.kt:
```diff
@@ -131,9 +104,8 @@ internal constructor(
-          // Registry state (note reverse order as saved above)
-          val owner = ViewTreeSavedStateRegistryOwner.get(oldViewMaybe) as KeyedStateRegistryOwner
-          registryHolder.saveRegistryController(savedKey, owner.controller)
+          // Save registry state (note reverse order as saved above)
```
Me either. :) I think this might be a vestige of when this method was more complex, will clean up.

rjrjr commented on workflow-ui/container-android/src/androidTest/java/com/squareup/workflow1/ui/backstack/test/ViewStateCacheTest.kt:
```diff
@@ -66,7 +66,9 @@ internal class ViewStateCacheTest {
     firstView.viewState = "hello world"
     // Cache expects there to be a registry owner on old views.
-    KeyedStateRegistryOwner.installAsSavedStateRegistryOwnerOn(firstView, "first")
+    StateRegistryAggregator({}) {}.installChildSavedStateRegistryControllerOn(
```
More temprorary cruft that I forgot to delete -- those are the same two `onSave / onRestore` params you asked about.

steve-the-edwards commented on workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt:
```diff
@@ -0,0 +1,300 @@
+  /**
+   * Used to observe the parent registry's lifecycle to know when it becomes `CREATED`
+   * and is ready for us to restore ourselves. This observer is only registered between
+   * calls to [attachToParentRegistry] and [detachFromParentRegistry], and will only
```
Is this fragile in the sense that if we called `setContentView` from _within_ `onCreate()` which would attach the view to the window, which would `attachToParentRegistry` we could miss the `ON_CREATE`?

I'm trying to think through how this works on config change and process restoration ... why is this working? Are we not calling `setCOntentView` via `onCreate`? Is this installed in `onCreate` but the 'ON_CREATE` lifecycle event happens after `onCreate` is complete?

steve-the-edwards commented on workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt:
```diff
@@ -0,0 +1,300 @@
+      )
+    }
+
+    // Even if the parent lifecycle is in a state further than CREATED, new observers are sent all
+    // the lifecycle events required to catch the observer up to the current state, so this handles
+    // both the cases where we're ready to immediately restore, and where we have to wait.
+    parentLifecycle.addObserver(lifecycleObserver)
```
Ahh. tada. this is the answer to my query above I guess. Good to know.

rjrjr replied:
It's SO handy. My single favorite thing about `Lifecycle`.

steve-the-edwards commented on workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt:
```diff
@@ -0,0 +1,300 @@
+  /**
+   * Puts a new [ViewTreeSavedStateRegistryOwner] in place on [view], registered
+   * with its [WorkflowLifecycleOwner]. **This method must be called before [view]
+   * is attached to a window.**
```
This _is_ quite the dance eh? SO how about them Fragments? (/joking!)

steve-the-edwards commented on workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt:
```diff
@@ -0,0 +1,300 @@
+   *  - all instances are dropped after [detachFromParentRegistry] is called
+   *
+   *  - if [view] is dropped but may be restored before [detachFromParentRegistry] is called,
+   *    as when pushing and popping a back stack, call [saveChildRegistryOwner] to capture
+   *    its state tear down its [ViewTreeSavedStateRegistryOwner]. The saved state
+   *    will be restored to the next [view] passed to [installChildSavedStateRegistryOwnerOn]
+   *    with the same [key]
```
Got lost in the grammar of the first sentence?  "call X to capture its state tear down its ..."?

rjrjr replied:
Is this better?

```
   *  - if [view] is dropped but may later be replaced with a new instance, as when
   *   pushing and popping a back stack, call [saveAndPruneChildRegistryOwner]
   *   to capture its state, and to stop requesting updates from it.
   *   The saved state will be restored to the next [view] passed to
   *   [installChildRegistryOwnerOn] with the same [key]
```

vgonda commented:
> 

😭

steve-the-edwards commented on workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt:
```diff
@@ -0,0 +1,300 @@
+      throw IllegalArgumentException("$key is already in use, it cannot be used register $view")
+    }
+    ViewTreeSavedStateRegistryOwner.get(view)?.let {
+      "$view already has ViewTreeSavedStateRegistryOwner: $it"
```
Is this supposed to be an exception? Isn't it just a naked string here?

rjrjr replied:
Yup, need to add a test covering it.

steve-the-edwards commented on workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt:
```diff
@@ -0,0 +1,300 @@
+   * with the same [KeyedStateRegistryOwner.key].
+   *
+   * This method can only be called once per key per instance of this class.
+   * After a key has been restored, its data will be removed from this object.
```
Is this part important? Why do we remove it? IS that what AndroidX is doing as well?

rjrjr replied:
This is a lie, I deleted it. We actually do call this method more than once.

steve-the-edwards commented on workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt:
```diff
@@ -0,0 +1,300 @@
+  }
+
+  private inline fun doIfRestored(block: (MutableMap<String, Bundle>) -> Unit) {
+    states?.let(block)
```
Don't we have `isRestored` also? We are kind of duplicating that here.

rjrjr replied:
If we check `isRestored` here we'd be duplicating the `states == null` check instead.

helios175 commented on workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/WorkflowAndroidXSupport.kt:
```diff
@@ -34,10 +38,52 @@ public object WorkflowAndroidXSupport {
+  @OptIn(WorkflowUiExperimentalApi::class)
+  public fun createStateRegistryKeyForContainer(containerView: View): String {
+    val nameSuffix = containerView.getRendering<Any>()
+      ?.let { Compatible.keyFor(it) }?.let { "-$it" }
```
If `Compatible.keyFor(it)` returns something it won't have a trailing `"-"` right?

rjrjr replied:
Nice catch.

Starting to second guess whether this method is even useful. Zach put it in place before I made the `Compatible.keyFor` usage more consistent -- he originally was doing `is Named<*>` checks. I'm going to try deleting this.

helios175 commented:
There's a bunch of
```
com.squareup.workflow1.ui.compose.ComposeViewTreeIntegrationTest > composition_state_is_not_restored_after_screen_is_removed_from_backstack[test(AVD) - 5.0.2] FAILED 
	java.lang.IllegalStateException: You can consumeRestoredStateForKey only after super.onCreate of corresponding component
	at androidx.savedstate.SavedStateRegistry.consumeRestoredStateForKey(SavedStateRegistry.java:77)

com.squareup.workflow1.ui.compose.ComposeViewTreeIntegrationTest > composition_state_is_restored_after_config_change[test(AVD) - 5.0.2] FAILED 
	java.lang.IllegalStateException: You can consumeRestoredStateForKey only after super.onCreate of corresponding component
	at androidx.savedstate.SavedStateRegistry.consumeRestoredStateForKey(SavedStateRegistry.java:77)

com.squareup.workflow1.ui.compose.ComposeViewTreeIntegrationTest > composition_is_disposed_when_navigated_away_dispose_on_destroy_strategy[test(AVD) - 5.0.2] FAILED 
	java.lang.IllegalStateException: You can consumeRestoredStateForKey only after super.onCreate of corresponding component
	at androidx.savedstate.SavedStateRegistry.consumeRestoredStateForKey(SavedStateRegistry.java:77)
```
in the instrumentation tests. Misconfigured or wrong assumptions?

rjrjr commented on workflow-ui/container-android/src/main/java/com/squareup/workflow1/ui/backstack/ViewStateCache.kt:
```diff
@@ -131,9 +104,8 @@ internal constructor(
-          val owner = ViewTreeSavedStateRegistryOwner.get(oldViewMaybe) as KeyedStateRegistryOwner
-          registryHolder.saveRegistryController(savedKey, owner.controller)
+          // Save registry state (note reverse order as saved above)
+          stateRegistryAggregator.saveChildSavedStateRegistryOwner(savedKey)
```
Left over comment, no longer meaningful. Deleted.

rjrjr commented on workflow-ui/container-android/src/main/java/com/squareup/workflow1/ui/backstack/ViewStateCache.kt:
```diff
@@ -37,30 +39,7 @@ internal constructor(
-        }
-    }
-  )
+  private val stateRegistryAggregator = StateRegistryAggregator()
```
How about `WorkflowSavedStateRegistryAggregator`? More parallel to `WorkflowLifecycleOwner`.

rjrjr commented on samples/containers/android/src/main/java/com/squareup/sample/container/overviewdetail/OverviewDetailContainer.kt:
```diff
@@ -53,11 +53,13 @@ class OverviewDetailContainer(view: View) : LayoutRunner<OverviewDetailScreen> {
-      // Since we have two sibling backstacks, we need to give them each different
-      // SavedStateRegistry key prefixes.
-      val overviewViewEnvironment = viewEnvironment
-        .withBackStackStateKeyPrefix(OverviewBackStackKey) + (OverviewDetailConfig to Overview)
```
The latter. `OverviewBackStackKey` was a hacky hint that was read only by `BackStackContainer`. Now that we have general support tied to `Compatible.compatibilityKey`, we can use `Named` for this kind of issue.

rjrjr commented on workflow-ui/container-android/src/main/java/com/squareup/workflow1/ui/modal/ModalContainer.kt:
```diff
@@ -75,6 +75,13 @@ public abstract class ModalContainer<ModalRenderingT : Any> @JvmOverloads constr
             )
+            // Ensure that each dialog has its own ViewTreeSavedStateRegistryOwner,
+            // so views in each dialog layer don't clash with other layers.
+            ref.savedStateRegistryKey = Compatible.keyFor(newScreen, i.toString())
```
`buildDialog` is an abstract method. By passing the id through it we're relying on subclasses to handle it properly. I did it this way to have one less mistake they could make. WDYT?

steve-the-edwards replied:
Gotcha. I guess its a bit awkward either way. Anyway - this is fine enough :)

rjrjr replied:
FWIW:

```kotlin
          // This is the unique id we'll use to make stateRegistryAggregator put a
          // SavedStateRegistryOwner on the dialog's decorView, as required by Compose.
          // We put it in this sketchy lateInit field on DialogRef rather than passing
          // it through abstract buildDialog method so that subclasses can't screw it up.
          ref.savedStateRegistryKey = Compatible.keyFor(modal, i.toString())
```

rjrjr commented on workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/KeyedStateRegistryOwner.kt:
```diff
@@ -0,0 +1,33 @@
+ * To create an instance, call [StateRegistryAggregator.installChildSavedStateRegistryControllerOn].
+ *
+ * @param key The key used to save and restore this controller from a [SavedStateRegistry].
+ * @param lifecycleOwner The [LifecycleOwner] that will be delegated to by this instance. Note that
+ * [SavedStateRegistryOwner] extends [LifecycleOwner].
```
Yeah, I'm not sure that this is useful information (and I think it was a mistake that the androidx team made this choice, can't see this redundant lifecycle is useful, or even safe to rely on). I'll delete the comment.

rjrjr replied:
Actually, this _is_ kind of useful, if only to explain why the hell we're doing this at all. Rephrased it a bit:

```
 * @param lifecycleOwner The [LifecycleOwner] that will be delegated to by this instance.
 * (Required because [SavedStateRegistryOwner] extends [LifecycleOwner] for no clear reason.)
```

rjrjr commented on workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/KeyedStateRegistryOwner.kt:
```diff
@@ -0,0 +1,33 @@
+ * [SavedStateRegistryOwner] extends [LifecycleOwner].
+ */
+@WorkflowUiExperimentalApi
+internal class KeyedStateRegistryOwner internal constructor(
```
Done

rjrjr commented on workflow-ui/container-android/src/main/java/com/squareup/workflow1/ui/backstack/BackStackContainer.kt:
```diff
@@ -47,17 +46,11 @@ import com.squareup.workflow1.ui.start
- * [BackStackContainer] is added with our old ID, that container will overwrite our state. Since
- * they'd both be using the same key, [SavedStateRegistry] would throw an exception. As long as this
- * container is detached before its ID is changed, it shouldn't be a problem.
+ * requirement, [BackStackContainer] tries to generate a best-effort unique key by using
```
stale

rjrjr commented on workflow-ui/container-android/src/main/java/com/squareup/workflow1/ui/modal/ModalContainer.kt:
```diff
@@ -48,24 +47,25 @@ public abstract class ModalContainer<ModalRenderingT : Any> @JvmOverloads constr
 
+  /**
+   * Provides a `ViewTreeSavedStateRegistryOwner` for each dialog,
+   * tied to that for this container view.
```
```kotlin
  /**
   * Provides a new `ViewTreeSavedStateRegistryOwner` for each dialog,
   * which will save to `ViewTreeSavedStateRegistryOwner` of this container view.
   */
  private val stateRegistryAggregator = WorkflowSavedStateRegistryAggregator()
```

rjrjr commented:
> There's a bunch of
> 
> ```
> com.squareup.workflow1.ui.compose.ComposeViewTreeIntegrationTest > composition_state_is_not_restored_after_screen_is_removed_from_backstack[test(AVD) - 5.0.2] FAILED 
> 	java.lang.IllegalStateException: You can consumeRestoredStateForKey only after super.onCreate of corresponding component
> 	at androidx.savedstate.SavedStateRegistry.consumeRestoredStateForKey(SavedStateRegistry.java:77)
> 
> com.squareup.workflow1.ui.compose.ComposeViewTreeIntegrationTest > composition_state_is_restored_after_config_change[test(AVD) - 5.0.2] FAILED 
> 	java.lang.IllegalStateException: You can consumeRestoredStateForKey only after super.onCreate of corresponding component
> 	at androidx.savedstate.SavedStateRegistry.consumeRestoredStateForKey(SavedStateRegistry.java:77)
> 
> com.squareup.workflow1.ui.compose.ComposeViewTreeIntegrationTest > composition_is_disposed_when_navigated_away_dispose_on_destroy_strategy[test(AVD) - 5.0.2] FAILED 
> 	java.lang.IllegalStateException: You can consumeRestoredStateForKey only after super.onCreate of corresponding component
> 	at androidx.savedstate.SavedStateRegistry.consumeRestoredStateForKey(SavedStateRegistry.java:77)
> ```
> 
> in the instrumentation tests. Misconfigured or wrong assumptions?

The caught me in a couple of actual bugs on the third SHA. Phew!

rjrjr commented on workflow-ui/compose/src/androidTest/java/com/squareup/workflow1/ui/compose/ComposeViewTreeIntegrationTest.kt:
```diff
@@ -436,13 +446,117 @@ internal class ComposeViewTreeIntegrationTest {
+        "Counter[$layer][$screen]: $counter",
+        Modifier
+          .clickable { counter++ }
+          .testTag("[$layer][$screen]")
```
Wow, nice catch. Thanks.

rjrjr commented on workflow-ui/container-android/src/androidTest/java/com/squareup/workflow1/ui/modal/test/ModalViewContainerLifecycleActivity.kt:
```diff
@@ -45,7 +45,11 @@ internal class ModalViewContainerLifecycleActivity : AbstractLifecycleTestActivi
     }
 
-    data class RecurseRendering(val wrapped: LeafRendering) : TestRendering()
+    data class RecurseRendering(
```
Unneeded

rjrjr commented:
Cool, cool, this passes locally:

```
* What went wrong:
Execution failed for task ':workflow-ui:container-android:ktlintAndroidTestSourceSetCheck'.
> A failure occurred while executing org.jlleitschuh.gradle.ktlint.worker.ConsoleReportWorkAction
   > KtLint found code style violations. Please see the following reports:
     - /home/runner/work/workflow-kotlin/workflow-kotlin/workflow-ui/container-android/build/reports/ktlint/ktlintAndroidTestSourceSetCheck/ktlintAndroidTestSourceSetCheck.json
```

rjrjr commented:
> Cool, cool, this passes locally:
> 
> ```
> * What went wrong:
> Execution failed for task ':workflow-ui:container-android:ktlintAndroidTestSourceSetCheck'.
> > A failure occurred while executing org.jlleitschuh.gradle.ktlint.worker.ConsoleReportWorkAction
>    > KtLint found code style violations. Please see the following reports:
>      - /home/runner/work/workflow-kotlin/workflow-kotlin/workflow-ui/container-android/build/reports/ktlint/ktlintAndroidTestSourceSetCheck/ktlintAndroidTestSourceSetCheck.json
> ```

Oh, b/c I hadn't pushed. Ray tired.

rjrjr commented:
@helios175 @vgonda @steve-the-edwards @RBusarow Everything passes locally. This should be ready for a final look through if you're so inclined.

I believe I've addressed all the feedback. The broken tests revealed one or two real problems, but no drastic changes were required. I've renamed a bunch of classes and methods, and was able to delete `createStateRegistryKeyForContainer()`, which is nice.

rjrjr commented:
Yay, I think I added a flaky test:

```
com.squareup.sample.TicTacToeEspressoTest > fullJourney[test(AVD) - 7.0] FAILED 
	androidx.test.espresso.NoMatchingViewException: No views in hierarchy found matching: with id is <com.squareup.sample.tictacworkflow:id/login_email>
```

rjrjr commented:
Publishing this branch as `1.6.0-SSR3-SNAPSHOT` while I do one more round of internal smoke testing.

rjrjr commented:
Maybe the answer is to add a non-optional `findParentLifecycle` function to `buildView`?

rjrjr commented:
Or maybe it's just fine to build it in. The only place we customize the `findParentLifecycle` function is when dialog builders apply it to `Dialog.window.decorView`, otherwise it's always the default function that looks in the view's parent. Redundant calls will just clobber each other with an identical function.

Wondering if there would be any remaining uses of `ViewStarter` with this change.

rjrjr commented:
> Wondering if there would be any remaining uses of `ViewStarter` with this change.

Yes, many, via `DecorativeViewFactory`.

rjrjr commented:
Going to track this separately, #673

steve-the-edwards commented on workflow-ui/container-android/src/main/java/com/squareup/workflow1/ui/modal/ModalContainer.kt:
```diff
@@ -167,12 +200,25 @@ public abstract class ModalContainer<ModalRenderingT : Any> @JvmOverloads constr
+     */
+    internal lateinit var savedStateRegistryKey: String
+
+    public fun copy(
+      modalRendering: ModalRenderingT,
+      viewEnvironment: ViewEnvironment,
+    ): DialogRef<ModalRenderingT> = DialogRef(modalRendering, viewEnvironment, dialog, extra).also {
```
Are we trying to enforce any behaviour here (i.e. specify `modalRendering` and `viewEnvironment` when making a new copy instance?). Otherwise we could just do a typical copy function like:

```
copy(
  modalRendering: ModalRenderingT = this.modalRendering,
  viewEnvironment: ViewEnvironment = this.viewEnvironment,
  dialog: Dialog = this.dialog,
  extra: Any? = this.extra,
```

rjrjr replied:
It won't matter in practice b/c this class is eliminated by the overhaul, but yeah, good point. I'll make it normal.

rjrjr replied:
Also note that that class itself is actually `protected`, so all of these `public`s are mistakes.

rjrjr replied:
Oh, actually required b/c it's no longer a `data` class.

steve-the-edwards reviewed (approved):
Still LGTM

rjrjr commented:
Internal tests are green. I'm making the `DialogRef.copy()` fix that @steve-the-edwards suggested, merging this, and then I'll publish 1.6.0.

rjrjr commented:
God damn it:

```kotlin
com.squareup.workflow1.ui.compose.ComposeViewTreeIntegrationTest > composition_is_restored_in_multiple_modals_backstacks_after_config_change[test(AVD) - 10] FAILED 
	java.lang.AssertionError: Failed to assert the following: (Text + EditableText = [Counter[0][0]: 1])
	The node is no longer in the tree, last known semantics:
```

rjrjr commented:
These are green at SHA 6e6bbf7e

rjrjr commented:
@RBusarow Looks like the only interesting change since the green build is #648, the move to the version catalogs. Yaaaay. No specific version change there is jumping out at me yet. Tweaked a few, no improvement.

rjrjr commented on workflow-ui/container-android/src/main/java/com/squareup/workflow1/ui/modal/ModalContainer.kt:
```diff
@@ -132,9 +153,21 @@ public abstract class ModalContainer<ModalRenderingT : Any> @JvmOverloads constr
+
+  override fun onAttachedToWindow() {
+    super.onAttachedToWindow()
+    val parentRegistry = stateRegistryOwnerFromViewTreeOrContext(this)
```
DRY. #681
